arangod/Aql/SubqueryStartExecutor.cpp

```diff
@@ -0,0 +1,85 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2019 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Michael Hackstein
+////////////////////////////////////////////////////////////////////////////////
+
+#include "SubqueryStartExecutor.h"
+
+#include "Aql/ExecutionState.h"
+#include "Aql/OutputAqlItemRow.h"
+#include "Aql/SingleRowFetcher.h"
+#include "Aql/Stats.h"
+
+using namespace arangodb;
+using namespace arangodb::aql;
+
+SubqueryStartExecutor::SubqueryStartExecutor(Fetcher& fetcher, Infos& infos)
+    : _fetcher(fetcher),
+      _state(ExecutionState::HASMORE),
+      _input(CreateInvalidInputRowHint{}) {}
+SubqueryStartExecutor::~SubqueryStartExecutor() = default;
+
+std::pair<ExecutionState, NoStats> SubqueryStartExecutor::produceRows(OutputAqlItemRow& output) {
+  while (!output.isFull()) {
+    TRI_ASSERT(!output.produced());
+    if (_state == ExecutionState::DONE && !_input.isInitialized()) {
+      // We need to handle shadowRows now. It is the job of this node to
+      // increase the shadow row depth
+      ShadowAqlItemRow shadowRow{CreateInvalidShadowRowHint{}};
+      std::tie(_state, shadowRow) = _fetcher.fetchShadowRow(output.numRowsLeft() / 2);
```

Comment:
Is this an exponential backoff?

Reply:
No, this is an upper bound.
For every input row this block will produce either 1 or 2 rows.
1 on shadow rows (pass through)
and 2 on data rows (passthrough + new shadow row)
Hence I only fetch half of what I require from upstream.

Reply:
i was wrong here, will remove